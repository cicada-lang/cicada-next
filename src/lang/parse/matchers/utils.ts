import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
