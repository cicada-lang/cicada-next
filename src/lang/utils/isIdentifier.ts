const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
