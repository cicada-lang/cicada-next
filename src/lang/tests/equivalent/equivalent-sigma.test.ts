import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Sigma", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent Type {
    exists (A: Type) A
  = exists (B: Type) B
  = exists (B: Type) id(Type, B)
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Sigma -- nested", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent Type {
    exists (A: Type, B: Type) Pair(A, B)
  = exists (B: Type, A: Type) Pair(B, A)
  = exists (B: Type, A: Type) Pair(id(Type, B), id(Type, A))
  = exists (B: Type, A: id(Type, Type)) Pair(id(Type, B), id(Type, A))
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Sigma -- fail", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    exists (A: Type, B: Type) Pair(A, B)
  = exists (A: Type, B: Type) Pair(B, A)
}

`)
})
