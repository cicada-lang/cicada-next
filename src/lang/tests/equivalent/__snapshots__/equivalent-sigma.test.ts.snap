// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`equivalent Sigma -- nested 1`] = `"refl(implicit Type, implicit exists (Type, Type, A) B): Equal(Type, exists (Type, Type, A) B, exists (Type, Type, B) A)"`;

exports[`equivalent Sigma 1`] = `"refl(implicit Type, implicit exists (Type) A): Equal(Type, exists (Type) A, exists (Type) B)"`;
